web-app/src/main/java/com/google/sps/data/QuadTree.java

```diff
@@ -59,4 +58,76 @@ public void printTree() {
     }
     System.out.printf("%n");
   }
+
+  public void insert(PoliceReport report) {
+    Node currentNode = root;
+    double reportLat = report.getLat();
+    double reportLng = report.getLng();
+
+    // Find a leaf to insert the report in
+    while (!currentNode.leaf) {
+      currentNode.numReports += 1;
+      for (Node child : currentNode.children) {
+        if (child.coordinates.inRectangle(reportLat, reportLng)) {
+          currentNode = child;
+          break;
+        }
+      }
+    }
+
+    currentNode.numReports += 1;
+    currentNode.reports.add(report);
+    // If max capacity has been exceeded, create child nodes
+    if (currentNode.numReports > reportCapacity && currentNode.depth < maxDepth) {
+      currentNode.children = reallocateReports(currentNode.reports, currentNode.coordinates, currentNode.depth);
+      currentNode.leaf = false;
+      currentNode.reports = null;
+    }
+  }
+
+  public Node[] reallocateReports(ArrayList<PoliceReport> reports, Rectangle coordinates, int depth) {
```

Comment:
I'm not sure if this should be public? I don't the client of the API is supposed to care about reallocating reports?